File: Maximize the Number of Target Nodes After Connecting Trees I/Maximize the Number of Target Nodes After Connecting Trees I.py
# Solution for Maximize the Number of Target Nodes After Connecting Trees I in PY

class Solution:
    def maxTargetNodes(self, edges1: List[List[int]], edges2: List[List[int]], k: int) -> List[int]:
        targetsFromTree1 = self.computeTargetNodesPerNode(edges1, k)
        targetsFromTree2 = self.computeTargetNodesPerNode(edges2, k - 1)
        maxTargetsInTree2 = max(targetsFromTree2)
        return [x + maxTargetsInTree2 for x in targetsFromTree1]

    def computeTargetNodesPerNode(self, edges: List[List[int]], maxDist: int) -> List[int]:
        totalNodes = len(edges) + 1
        adj = defaultdict(list)
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
        return [self.BFS(i, adj, maxDist, totalNodes) for i in range(totalNodes)]

    def BFS(self, startNode: int, adj: dict, maxDist: int, totalNodes: int) -> int:
        q = deque([(startNode, 0)])
        visited = [False] * totalNodes
        visited[startNode] = True
        count = 0
        while q:
            node, dist = q.popleft()
            if dist > maxDist:
                continue
            count += 1
            for neighbor in adj[node]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    q.append((neighbor, dist + 1))
        return count

File: Maximize the Number of Target Nodes After Connecting Trees I/Maximize the Number of Target Nodes After Connecting Trees I.java
// Solution for Maximize the Number of Target Nodes After Connecting Trees I in JAVA

class Solution {
    public int[] maxTargetNodes(int[][] edges1, int[][] edges2, int k) {
        int[] targetsFromTree1 = computeTargetNodesPerNode(edges1, k);
        int[] targetsFromTree2 = computeTargetNodesPerNode(edges2, k - 1);
        int maxTargetsInTree2 = 0;
        for (int val : targetsFromTree2) maxTargetsInTree2 = Math.max(maxTargetsInTree2, val);
        for (int i = 0; i < targetsFromTree1.length; i++) targetsFromTree1[i] += maxTargetsInTree2;
        return targetsFromTree1;
    }

    private int[] computeTargetNodesPerNode(int[][] edges, int maxDist) {
        int totalNodes = edges.length + 1;
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int[] edge : edges) {
            adj.computeIfAbsent(edge[0], x -> new ArrayList<>()).add(edge[1]);
            adj.computeIfAbsent(edge[1], x -> new ArrayList<>()).add(edge[0]);
        }
        int[] nodeTargetCounts = new int[totalNodes];
        for (int i = 0; i < totalNodes; i++) nodeTargetCounts[i] = BFS(i, adj, maxDist, totalNodes);
        return nodeTargetCounts;
    }

    private int BFS(int startNode, Map<Integer, List<Integer>> adj, int maxDist, int totalNodes) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{startNode, 0});
        boolean[] visited = new boolean[totalNodes];
        visited[startNode] = true;
        int count = 0;
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int node = cur[0], dist = cur[1];
            if (dist > maxDist) continue;
            count++;
            for (int neighbor : adj.getOrDefault(node, new ArrayList<>())) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.offer(new int[]{neighbor, dist + 1});
                }
            }
        }
        return count;
    }
}
